chains/solana/contracts/programs/ccip-router/src/fee_quoter.rs

```diff
@@ -4,42 +4,143 @@ use ethnum::U256;
+    let fee_token_price = get_validated_token_price(fee_token_config)?;
     let _packed_gas_price = get_validated_gas_price(dest_chain)?;
 
+    let network_fee = network_fee(
+        message,
+        dest_chain,
+        additional_token_configs,
+        additional_token_configs_for_dest_chain,
+    )?;
+
     // TODO un-hardcode
-    let network_fee = U256::new(1);
-    let execution_cost = U256::new(1);
-    let data_availability_cost = U256::new(1);
+    let execution_cost = 1u32.e(18);
+    let data_availability_cost = 1u32.e(18);
 
-    let amount = (network_fee + execution_cost + data_availability_cost) / token_price;
+    let premium_multiplier = U256::new(fee_token_config.premium_multiplier_wei_per_eth.into());
+    let amount =
+        (network_fee.premium.0 * premium_multiplier + execution_cost + data_availability_cost)
+            / fee_token_price.0;
     let amount: u64 = amount
         .try_into()
         .map_err(|_| CcipRouterError::InvalidTokenPrice)?;
 
     Ok(SolanaTokenAmount { amount, token })
 }
 
+struct NetworkFee {
+    premium: Usd18Decimals,
+    _token_transfer_gas: U256,
+    _token_transfer_bytes_overhead: U256,
+}
+
+fn network_fee(
+    message: &Solana2AnyMessage,
+    dest_chain: &DestChain,
+    token_configs: &[BillingTokenConfig],
+    token_configs_for_dest_chain: &[PerChainPerTokenConfig],
+) -> Result<NetworkFee> {
+    if message.token_amounts.is_empty() {
+        return Ok(NetworkFee {
+            premium: Usd18Decimals::from_usd_cents(dest_chain.config.network_fee_usdcents),
+            _token_transfer_gas: U256::ZERO,
+            _token_transfer_bytes_overhead: U256::ZERO,
+        });
+    }
+
+    let mut premium = Usd18Decimals::default();
+    let mut token_transfer_gas = U256::ZERO;
+    let mut token_transfer_bytes_overhead = U256::ZERO;
+
+    for token_amount in &message.token_amounts {
+        let config_for_dest_chain = token_configs_for_dest_chain
+            .iter()
+            .find(|c| c.mint == token_amount.token)
+            .ok_or(CcipRouterError::UnsupportedToken)?;
+
+        let (token_fee, token_gas, token_overhead) = if config_for_dest_chain.billing.is_enabled {
+            token_network_fees(token_configs, token_amount, config_for_dest_chain)?
+        } else {
+            // If the token has no specific overrides configured, we use the global defaults.
+            global_network_fees(dest_chain)
+        };
+
+        premium += token_fee;
+        token_transfer_gas += token_gas;
+        token_transfer_bytes_overhead += token_overhead;
+    }
+
+    Ok(NetworkFee {
+        premium,
+        _token_transfer_gas: token_transfer_gas,
+        _token_transfer_bytes_overhead: token_transfer_bytes_overhead,
+    })
+}
+
+fn token_network_fees(
+    token_configs: &[BillingTokenConfig],
+    token_amount: &SolanaTokenAmount,
+    config_for_dest_chain: &PerChainPerTokenConfig,
+) -> Result<(Usd18Decimals, U256, U256)> {
```

Comment:
Should this return `Result<NetworkFee>` instead?

Reply:
Yes, that's a slip up, good eye 😅